pkg/controllers/overrider/common_integration_test.go

```diff
@@ -0,0 +1,277 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package overrider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	fleetv1alpha1 "go.goms.io/fleet/apis/placement/v1alpha1"
+	fleetv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+	"go.goms.io/fleet/pkg/utils/controller"
+	"go.goms.io/fleet/pkg/utils/labels"
+)
+
+const (
+	eventuallyTimeout    = time.Second * 10
+	consistentlyDuration = time.Second * 5
+	interval             = time.Millisecond * 250
+)
+
+func getClusterResourceOverrideSpec() fleetv1alpha1.ClusterResourceOverrideSpec {
+	return fleetv1alpha1.ClusterResourceOverrideSpec{
+		ClusterResourceSelectors: []fleetv1beta1.ClusterResourceSelector{
+			{
+				Group:   "",
+				Version: "v1",
+				Kind:    "Pod",
+			},
+		},
+		Policy: &fleetv1alpha1.OverridePolicy{
+			OverrideRules: []fleetv1alpha1.OverrideRule{
+				{
+					JSONPatchOverrides: []fleetv1alpha1.JSONPatchOverride{
+						{
+							Operator: fleetv1alpha1.JSONPatchOverrideOpReplace,
+							Path:     "spec.replica",
+							Value:    "3",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func getClusterResourceOverride(testOverrideName string) *fleetv1alpha1.ClusterResourceOverride {
+	return &fleetv1alpha1.ClusterResourceOverride{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:       testOverrideName,
+			Finalizers: []string{fleetv1alpha1.OverrideFinalizer},
+		},
+		Spec: getClusterResourceOverrideSpec(),
+	}
+}
+
+func getClusterResourceOverrideSnapshot(testOverrideName string, index int) *fleetv1alpha1.ClusterResourceOverrideSnapshot {
+	return &fleetv1alpha1.ClusterResourceOverrideSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf(fleetv1alpha1.OverrideSnapshotNameFmt, testOverrideName, index),
+			Labels: map[string]string{
+				fleetv1alpha1.OverrideIndexLabel:    strconv.Itoa(index),
+				fleetv1beta1.IsLatestSnapshotLabel:  "true",
+				fleetv1alpha1.OverrideTrackingLabel: testOverrideName,
+			},
+		},
+		Spec: fleetv1alpha1.ClusterResourceOverrideSnapshotSpec{
+			OverrideHash: []byte("hash"),
+			OverrideSpec: getClusterResourceOverrideSpec(),
+		},
+	}
+}
+
+var _ = Describe("Test ClusterResourceOverride common logic", func() {
```

Comment:
The integration test looks weird as it manually triggered some of the funcs, instead of by creating/update/deleting the override and then validating the controller behavior.

It looks like unit tests.

Reply:
yep, it's a UT with a real API server. The IT was for the whole reconciler.

Reply:
I will see if renaming it as UT will break the test.